apps/uvc-app/uvc-gadget.cpp

```diff
@@ -241,760 +236,136 @@ struct uvc_device {
+    int ret;
+    struct v4l2_buffer ubuf;
 
-        ret = ioctl(v4l2_subdev_fd, VIDIOC_S_EXT_CTRLS, &extCtrls);
-        if (ret < 0) {
-            fprintf(stderr, "Error in Programming AFE\n");
-        }
+    if (!DeviceStartedStreaming) {
+        return 0;
     }
-    else { /* afe programming data to be read from file */
-        fp = fopen( v4l2_subdev_prog_file, "rb"); 
-        if (fp!= NULL) {
-            fseek(fp, 0L, SEEK_END);
-            flen = ftell(fp);  
-            fseek(fp, 0L, SEEK_SET);
-
-            while (written_bytes < flen) 
-            {
-                fseek(fp, written_bytes, 0);
-
-                extCtrl.size = 2048 * sizeof (unsigned short);
-                extCtrl.p_u16 = (unsigned short *) buf;
-                extCtrl.id = V4L2_CID_AD_DEV_SET_CHIP_CONFIG;
-
-                if ((flen - written_bytes) > max_write_size) {
-                    fread(buf, 1, max_write_size, fp);
-
-                    memset(&extCtrls, 0, sizeof(struct v4l2_ext_controls));
-                    extCtrls.controls = &extCtrl;
-                    extCtrls.count = 1;
-                    extCtrls.ctrl_class = 0x00a00000;
-
-                    ret = ioctl(v4l2_subdev_fd, VIDIOC_S_EXT_CTRLS, &extCtrls);
-                    if (ret < 0) {
-                        fprintf(stderr, "Error in Programming AFE\n");
-                        break;
-                    }
 
-                    written_bytes += max_write_size;
-                }
-                else {
-                    memset(buf, 0, max_write_size);
-                    fread(buf, 1, (flen - written_bytes), fp);
-
-                    extCtrl.size = 2048 * sizeof (unsigned short);
-                    extCtrl.p_u16 = (unsigned short *) buf;
-                    extCtrl.id = V4L2_CID_AD_DEV_SET_CHIP_CONFIG;
-
-                    memset(&extCtrls, 0, sizeof(struct v4l2_ext_controls));
-                    extCtrls.controls = &extCtrl;
-                    extCtrls.count = 1;
-                    extCtrls.ctrl_class = 0x00a00000;
-
-                    ret = ioctl(v4l2_subdev_fd, VIDIOC_S_EXT_CTRLS, &extCtrls);
-                    if (ret < 0) {
-                        fprintf(stderr, "Error in Programming AFE\n");
-                        break;
-                    }
-
-                    written_bytes = flen;
-                    fclose(fp);
-                }
-            }
-        }
-        else {
-            printf ("File cannot be Found or Opened \n");
-        }
-    }
+    /* Queue video buffer to UVC domain. */
+    CLEAR(ubuf);
 
-    printf ("<< v4l2_subdev_program\n");
+    ubuf.type = V4L2_BUF_TYPE_VIDEO_OUTPUT;
 
-    return ret;
-}
+    ubuf.memory = V4L2_MEMORY_USERPTR;
 
-static void v4l2_subdev_close()
-{
-    close (v4l2_subdev_fd);
-}
+    if (!udev->hasNewFrame) {
+        return 0;
+    }
 
-static int readAfeTemp(float* temperature) {
-    temp_sensor tdev;
+    auto status = static_cast<LocalDevice *>(device)->dequeueBuffer(
```

Comment:
The safely downcast use dynamic_cast

Reply:
I've changed the code to use `dynamic_cast`.